### DISCUSSION TITLE
DatasyncClient Singleton

### DISCUSSION THREAD

**Discussion by PabloRuiz-TT** [Azure/azure-mobile-apps#772] [Q&A] [answered]
I have created a class that allows me to get the tables of my application, making it generic and thus only have a configuration of the datasyn client, then I pass it in the MauiProgram as a singleton.
Is this correct, any suggestions ?

`public class SaygerDbContext<TEntity> where TEntity : class
{
    private DatasyncClient _client = null;

    private IOfflineTable<TEntity> _table = null;

    private bool _initialized = false;

    private readonly SemaphoreSlim _asyncLock = new(1, 1);

    private async Task InitializeAsync()
    {
        if (_initialized) return;

        try
        {
            await _asyncLock.WaitAsync();

            if (_initialized) return;

            string connectionString = new UriBuilder()
            {
                Scheme = "file",
                Path = $"{FileSystem.AppDataDirectory}/sayger.db",
                Query = "?mode=rwc"
            }.ToString();

            OfflineSQLiteStore store = new(connectionString);

            store.DefineTable<TEntity>();

            var options = new DatasyncClientOptions
            {
                HttpPipeline = new HttpMessageHandler[] { new LoggingHandler() },
                OfflineStore = store
            };

            _client = new DatasyncClient(Constants.Uri, options);

            await _client.InitializeOfflineStoreAsync();

            _table = _client.GetOfflineTable<TEntity>();

            _initialized = true;

        }
        catch (Exception)
        {
        }

        finally
        {
            _asyncLock.Release();
        }
    }

    public async Task<IOfflineTable<TEntity>> GetOfflineTableAsync()
    {
        await InitializeAsync();

        return _table;
    }`

**Comment by Arslan007** [comment]
I think it will not work with Generic type <T>. You have to initialize all the Tables at once and then call `await _client.InitializeOfflineStoreAsync();` otherwise you will get error as first table will be initialized and rest will be not.

**Reply by adrianhall** [reply]
Agreed - you have multiple tables; being generic is not useful, and if you use it generically, you will get a conflict at the offline store because two different tables are trying to use the same offline store at the same time.

Better to create a "ServiceClient" class that has all the tables defined and ready in it, then have something like:

```csharp
public IOfflineTable<TEntity> GetTable<TEntity>() where TEntity : DatasyncClientData 
=> _client.GetOfflineTable<TEntity>();
```

... to make getting the tables abstracted from the client.  You may even want to wrap the IOfflineTable<TEntity> in your own class so that you can do access logging or whatever else you want, and further encapsulate the Azure Mobile Apps code and hide it from your application.

**Reply by PabloRuiz-TT** [reply, marked as answer]
Thank you for your answers, based on the above, it is correct as I understood and expressed. Example: 

```csharp
public class SaygerDbContext
{
    private DatasyncClient _client = null;

    private bool _initialized = false;

    private readonly SemaphoreSlim _asyncLock = new(1, 1);

    private async Task InitializeAsync()
    {
        if (_initialized) return;

        try
        {
            await _asyncLock.WaitAsync();

            if (_initialized) return;

            string connectionString = new UriBuilder()
            {
                Scheme = "file",
                Path = $"{FileSystem.AppDataDirectory}/sayger.db",
                Query = "?mode=rwc"
            }.ToString();

            OfflineSQLiteStore store = new(connectionString);

            store.DefineTable<Branch>();

            store.DefineTable<Company>();

            store.DefineTable<SalesRep>();

            var options = new DatasyncClientOptions
            {
                HttpPipeline = new HttpMessageHandler[] { new LoggingHandler() },
                OfflineStore = store
            };

            _client = new DatasyncClient(Constants.Uri, options);

            await _client.InitializeOfflineStoreAsync();

            _initialized = true;

        }
        catch (Exception)
        {
        }

        finally
        {
            _asyncLock.Release();
        }
    }


    public async Task<IOfflineTable<TEntity>> GetTable<TEntity>() where TEntity : DatasyncClientData
    {
        await InitializeAsync();

        return _client.GetOfflineTable<TEntity>();
    }
```

**Reply by adrianhall** [reply]
Yep - this is a basic copy of the tutorial code.
